Laboratorio 2 - Deep Learning
Mariana David 201055


En parejas, utilicen el dataset publicado para definir una implementación de redes neuronales. Tienen libertad
de
• Elegir el objetivo de la red, puede ser clasificación (genero) o regresión (aprobación, ganancias).
• Elegir el subconjunto variables a utilizar.
Una vez elegido el objetivo y los campos a usar en la red construyan 3 redes diferentes donde,
• Se utilicen diferentes funciones de activación.
• Diferentes números de capas y neuronas.
• Diferentes técnicas de regularización.
Reporte.
1. Describa el objetivo de la red y por qué considera que la implementación de una red neuronal puede
obtener el resultado deseado.
2. Describa la composición y los resultados obtenidos por red neuronal.
3. Discuta la diferencia de rendimiento y conceptuales en la composición y resultados obtenidos en cada
red neural.
4. Seleccione la red neuronal optima y justifique su respuesta.
Pueden utilizar como base el archivo bn_implementation.ipynb publicado en la plataforma.

Importaciones


In [1]:
#import tensorflow as tf
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, BatchNormalization, LeakyReLU, Activation
from tensorflow.keras.optimizers import RMSprop, Adagrad, Adam

DataSet Seleccionado


In [8]:
dataset = "movie_statistic_dataset.csv" # Importacion del dataset 
df = pd.read_csv(dataset) # Carga del dataset
df # Muestra del dataset.
df.columns #Muestra de los títulos del dataset
df.head() #Mostrar data

movie_title production_date  \
0                     Avatar: The Way of Water      2022-12-09   
1                            Avengers: Endgame      2019-04-23   
2  Pirates of the Caribbean: On Stranger Tides      2011-05-20   
3                      Avengers: Age of Ultron      2015-04-22   
4                       Avengers: Infinity War      2018-04-25   

                     genres  runtime_minutes  director_name  \
0  Action,Adventure,Fantasy            192.0  James Cameron   
1    Action,Adventure,Drama            181.0              -   
2  Action,Adventure,Fantasy            137.0   Rob Marshall   
3   Action,Adventure,Sci-Fi            141.0    Joss Whedon   
4   Action,Adventure,Sci-Fi            149.0              -   

              director_professions director_birthYear director_deathYear  \
0         writer,producer,director               1954              alive   
1                                -                  -                  -   
2  director,miscellaneous,producer               1960              alive   
3         writer,producer,director               1964              alive   
4                                -                  -                  -   

   movie_averageRating  movie_numerOfVotes  approval_Index  \
0                  7.8            277543.0        7.061101   
1                  8.4           1143642.0        8.489533   
2                  6.6            533763.0        6.272064   
3                  7.3            870573.0        7.214013   
4                  8.4           1091968.0        8.460958   

   Production budget $  Domestic gross $  Worldwide gross $  
0            460000000         667830256         2265935552  
1            400000000         858373000         2794731755  
2            379000000         241071802         1045713802  
3            365000000         459005868         1395316979  
4            300000000         678815482         2048359754

***OBJETIVO:*** el objetivo es predecir la aprobación (approval_Index) de una película en función de las variables disponibles en el dataset.

In [2]:
# Importar las bibliotecas necesarias
import pandas as pd
import numpy as np
import tensorflow as tf
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error

# Cargar el dataset
data = pd.read_csv('movie_statistic_dataset.csv')  # Cambia esto al nombre de tu archivo

# Seleccionar las variables predictoras (características) y la variable de destino
features = ['runtime_minutes', 'movie_averageRating', 'movie_numerOfVotes', 'Production budget $', 'Domestic gross $', 'Worldwide gross $']
target = 'approval_Index'

# Dividir el conjunto de datos en características (X) y variable objetivo (y)
X = data[features].values
y = data[target].values

# Dividir los datos en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Normalizar las características
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Construir el modelo de red neuronal
model = tf.keras.models.Sequential([
    tf.keras.layers.Input(shape=(X_train_scaled.shape[1],)),
    tf.keras.layers.Dense(64, activation='relu'),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(1)
])

# Compilar el modelo
model.compile(optimizer='adam', loss='mean_squared_error')

# Entrenar el modelo
model.fit(X_train_scaled, y_train, epochs=50, batch_size=32, validation_split=0.1)

# Realizar predicciones en el conjunto de prueba
y_pred = model.predict(X_test_scaled)

# Calcular el RMSE en el conjunto de prueba
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print(f'RMSE en el conjunto de prueba: {rmse}')


Epoch 1/50
99/99 [==============================] - 2s 6ms/step - loss: 18.6144 - val_loss: 10.9701
Epoch 2/50
99/99 [==============================] - 0s 3ms/step - loss: 7.0924 - val_loss: 4.4535
Epoch 3/50
99/99 [==============================] - 0s 3ms/step - loss: 3.5398 - val_loss: 2.8337
Epoch 4/50
99/99 [==============================] - 0s 2ms/step - loss: 2.4419 - val_loss: 1.9037
Epoch 5/50
99/99 [==============================] - 0s 3ms/step - loss: 1.7894 - val_loss: 1.3165
Epoch 6/50
99/99 [==============================] - 0s 3ms/step - loss: 1.3390 - val_loss: 0.9665
Epoch 7/50
99/99 [==============================] - 0s 3ms/step - loss: 1.0115 - val_loss: 0.7439
Epoch 8/50
99/99 [==============================] - 0s 4ms/step - loss: 0.8405 - val_loss: 0.6203
Epoch 9/50
99/99 [==============================] - 0s 3ms/step - loss: 0.7626 - val_loss: 0.5402
Epoch 10/50
99/99 [==============================] - 0s 3ms/step - loss: 0.6778 - val_loss: 0.4926
Epoch 11/50
99/99

Se analizara el Worlwide gross dependiendo del genero de la película 

In [6]:
# Seleccionamos las columnas que queremos combinar
selected_columns = ['Production budget $', 'Worldwide gross $']

# Utilizamos la función concat() para combinar las columnas en una sola variable
combined_variable = pd.concat([df[col] for col in selected_columns], axis=1)

# Renombramos las columnas del nuevo DataFrame resultante
combined_variable.columns = ['movie_title', 'genres']

# Ahora 'combined_varia(ble' contiene todas las columnas seleccionadas combinadas en una sola variable
print(combined_variable)

NameError: name 'df' is not defined

***Red Neuronal 1 - ReLU y Dropout:***

In [4]:
import pandas as pd
import numpy as np
import tensorflow as tf
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error

# Cargar el dataset
data = pd.read_csv('movie_statistic_dataset.csv')  # Cambia esto al nombre de tu archivo

# Seleccionar las variables predictoras (características) y la variable de destino
features = ['runtime_minutes', 'movie_averageRating', 'movie_numerOfVotes', 'Production budget $', 'Domestic gross $', 'Worldwide gross $']
target = 'approval_Index'

# Dividir el conjunto de datos en características (X) y variable objetivo (y)
X = data[features].values
y = data[target].values

# Dividir los datos en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Normalizar las características
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Construir el modelo de red neuronal
model_relu = tf.keras.models.Sequential([
    tf.keras.layers.Input(shape=(X_train_scaled.shape[1],)),
    tf.keras.layers.Dense(64, activation='relu'),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(1)
])

# Compilar el modelo
model_relu.compile(optimizer='adam', loss='mean_squared_error')

# Entrenar el modelo
model_relu.fit(X_train_scaled, y_train, epochs=50, batch_size=32, validation_split=0.1)

# Realizar predicciones en el conjunto de prueba
y_pred_relu = model_relu.predict(X_test_scaled)

# Calcular el RMSE en el conjunto de prueba
rmse_relu = np.sqrt(mean_squared_error(y_test, y_pred_relu))
print(f'RMSE en el conjunto de prueba (ReLU): {rmse_relu}')


Epoch 1/50
99/99 [==============================] - 1s 5ms/step - loss: 17.7013 - val_loss: 10.6120
Epoch 2/50
99/99 [==============================] - 0s 4ms/step - loss: 6.4147 - val_loss: 3.9026
Epoch 3/50
99/99 [==============================] - 0s 3ms/step - loss: 2.9835 - val_loss: 2.4471
Epoch 4/50
99/99 [==============================] - 0s 3ms/step - loss: 2.3553 - val_loss: 1.7732
Epoch 5/50
99/99 [==============================] - 0s 4ms/step - loss: 1.7082 - val_loss: 1.2860
Epoch 6/50
99/99 [==============================] - 0s 3ms/step - loss: 1.2611 - val_loss: 0.9904
Epoch 7/50
99/99 [==============================] - 0s 3ms/step - loss: 1.1074 - val_loss: 0.7929
Epoch 8/50
99/99 [==============================] - 0s 3ms/step - loss: 0.9271 - val_loss: 0.6216
Epoch 9/50
99/99 [==============================] - 0s 3ms/step - loss: 0.7558 - val_loss: 0.5554
Epoch 10/50
99/99 [==============================] - 0s 3ms/step - loss: 0.7147 - val_loss: 0.4996
Epoch 11/50
99/99

***Red Neuronal 2 - Tanh y L2 Regularization:***

In [7]:
# Construir el modelo de red neuronal
model_tanh = tf.keras.models.Sequential([
    tf.keras.layers.Input(shape=(X_train_scaled.shape[1],)),
    tf.keras.layers.Dense(128, activation='tanh'),
    tf.keras.layers.Dense(64, activation='tanh'),
    tf.keras.layers.Dense(1, kernel_regularizer=tf.keras.regularizers.l2(0.01))
])

# Compilar el modelo
model_tanh.compile(optimizer='adam', loss='mean_squared_error')

# Entrenar el modelo
model_tanh.fit(X_train_scaled, y_train, epochs=50, batch_size=32, validation_split=0.1)

# Realizar predicciones en el conjunto de prueba
y_pred_tanh = model_tanh.predict(X_test_scaled)

# Calcular el RMSE en el conjunto de prueba
rmse_tanh = np.sqrt(mean_squared_error(y_test, y_pred_tanh))
print(f'RMSE en el conjunto de prueba (Tanh): {rmse_tanh}')


Epoch 1/50
99/99 [==============================] - 2s 6ms/step - loss: 10.0341 - val_loss: 0.4193
Epoch 2/50
99/99 [==============================] - 0s 3ms/step - loss: 0.2939 - val_loss: 0.3321
Epoch 3/50
99/99 [==============================] - 0s 3ms/step - loss: 0.2644 - val_loss: 0.3146
Epoch 4/50
99/99 [==============================] - 0s 2ms/step - loss: 0.2483 - val_loss: 0.2982
Epoch 5/50
99/99 [==============================] - 0s 3ms/step - loss: 0.2381 - val_loss: 0.2862
Epoch 6/50
99/99 [==============================] - 0s 4ms/step - loss: 0.2276 - val_loss: 0.2841
Epoch 7/50
99/99 [==============================] - 0s 4ms/step - loss: 0.2218 - val_loss: 0.2737
Epoch 8/50
99/99 [==============================] - 0s 3ms/step - loss: 0.2123 - val_loss: 0.2621
Epoch 9/50
99/99 [==============================] - 0s 3ms/step - loss: 0.2082 - val_loss: 0.2629
Epoch 10/50
99/99 [==============================] - 0s 3ms/step - loss: 0.1999 - val_loss: 0.2671
Epoch 11/50
99/99 

***Red Neuronal 3 - Leaky ReLU y Batch Normalization:***

In [8]:
# Construir el modelo de red neuronal
model_leaky_relu = tf.keras.models.Sequential([
    tf.keras.layers.Input(shape=(X_train_scaled.shape[1],)),
    tf.keras.layers.Dense(256, activation=tf.keras.layers.LeakyReLU(alpha=0.01)),
    tf.keras.layers.Dense(128, activation=tf.keras.layers.LeakyReLU(alpha=0.01)),
    tf.keras.layers.Dense(64, activation=tf.keras.layers.LeakyReLU(alpha=0.01)),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Dense(1)
])

# Compilar el modelo
model_leaky_relu.compile(optimizer='adam', loss='mean_squared_error')

# Entrenar el modelo
model_leaky_relu.fit(X_train_scaled, y_train, epochs=50, batch_size=32, validation_split=0.1)

# Realizar predicciones en el conjunto de prueba
y_pred_leaky_relu = model_leaky_relu.predict(X_test_scaled)

# Calcular el RMSE en el conjunto de prueba
rmse_leaky_relu = np.sqrt(mean_squared_error(y_test, y_pred_leaky_relu))
print(f'RMSE en el conjunto de prueba (Leaky ReLU): {rmse_leaky_relu}')


Epoch 1/50
99/99 [==============================] - 2s 6ms/step - loss: 19.7531 - val_loss: 13.0949
Epoch 2/50
99/99 [==============================] - 0s 4ms/step - loss: 5.5527 - val_loss: 2.2094
Epoch 3/50
99/99 [==============================] - 0s 5ms/step - loss: 0.4285 - val_loss: 0.9329
Epoch 4/50
99/99 [==============================] - 0s 4ms/step - loss: 0.2192 - val_loss: 0.4581
Epoch 5/50
99/99 [==============================] - 0s 5ms/step - loss: 0.2120 - val_loss: 0.3116
Epoch 6/50
99/99 [==============================] - 0s 4ms/step - loss: 0.2216 - val_loss: 0.4198
Epoch 7/50
99/99 [==============================] - 0s 4ms/step - loss: 0.1996 - val_loss: 0.1436
Epoch 8/50
99/99 [==============================] - 0s 3ms/step - loss: 0.1972 - val_loss: 0.2513
Epoch 9/50
99/99 [==============================] - 0s 4ms/step - loss: 0.1764 - val_loss: 0.1400
Epoch 10/50
99/99 [==============================] - 0s 3ms/step - loss: 0.1800 - val_loss: 0.1490
Epoch 11/50
99/99

***Graficos de las 3 redes neuronales***

In [9]:
import matplotlib.pyplot as plt

# Entrenar el modelo1
history_relu = model_relu.fit(X_train_scaled, y_train, epochs=50, batch_size=32, validation_split=0.1, verbose=0)

# Graficar las curvas de entrenamiento y validación
plt.plot(history_relu.history['accuracy'], label='Train Accuracy')
plt.plot(history_relu.history['val_accuracy'], label='Validation Accuracy')
plt.xlabel('Epochs')
plt.ylabel('Accuracy')
plt.title('Red Neuronal 1 - ReLU y Dropout')
plt.legend()
plt.show()

# Entrenar el modelo2
history_tanh = model_tanh.fit(X_train_scaled, y_train, epochs=50, batch_size=32, validation_split=0.1, verbose=0)

# Graficar las curvas de entrenamiento y validación
plt.plot(history_tanh.history['accuracy'], label='Train Accuracy')
plt.plot(history_tanh.history['val_accuracy'], label='Validation Accuracy')
plt.xlabel('Epochs')
plt.ylabel('Accuracy')
plt.title('Red Neuronal 2 - Tanh y L2 Regularization')
plt.legend()
plt.show()

# Entrenar el modelo3
history_leaky_relu = model_leaky_relu.fit(X_train_scaled, y_train, epochs=50, batch_size=32, validation_split=0.1, verbose=0)

# Graficar las curvas de entrenamiento y validación
plt.plot(history_leaky_relu.history['accuracy'], label='Train Accuracy')
plt.plot(history_leaky_relu.history['val_accuracy'], label='Validation Accuracy')
plt.xlabel('Epochs')
plt.ylabel('Accuracy')
plt.title('Red Neuronal 3 - Leaky ReLU y Batch Normalization')
plt.legend()
plt.show()


KeyError: 'accuracy'

***REPORTE***

El objetivo es predecir la aprobación (approval_Index) de una película en función de las variables disponibles en el dataset. Para esto se utilizó diferentes configuraciones de redes neuronales para comparar su rendimiento.

**1. Objetivo de la red y justificación:**

El objetivo de la red neuronal es predecir la aprobación (approval_Index) de una película. Se considera que la implementación de una red neuronal puede obtener el resultado deseado debido a la capacidad de las redes neuronales para aprender relaciones complejas y no lineales entre las variables de entrada y la variable de salida. Además, dado que el dataset contiene múltiples características de las películas, una red neuronal puede capturar patrones y correlaciones que pueden no ser evidentes en enfoques de modelado más tradicionales.

**2. Composición y resultados de las redes neuronales:**

**Red Neuronal 1:**
- Función de activación: ReLU (Rectified Linear Unit).
- Capas: 3 capas (entrada, oculta, salida).
- Neuronas: 64 en la capa oculta.
- Técnica de regularización: Dropout con tasa 0.2.

Resultados: RMSE en conjunto de prueba: 0.15

**Red Neuronal 2:**
- Función de activación: Tanh (Tangente hiperbólica).
- Capas: 4 capas (entrada, oculta1, oculta2, salida).
- Neuronas: 128 en oculta1, 64 en oculta2.
- Técnica de regularización: L2 regularization.

Resultados: RMSE en conjunto de prueba: 0.17

**Red Neuronal 3:**
- Función de activación: Leaky ReLU.
- Capas: 5 capas (entrada, oculta1, oculta2, oculta3, salida).
- Neuronas: 256 en oculta1, 128 en oculta2, 64 en oculta3.
- Técnica de regularización: Batch normalization.

Resultados: RMSE en conjunto de prueba: 0.16

**3. Diferencias de rendimiento y conceptuales:**

- La Red Neuronal 1 utiliza la función de activación ReLU y Dropout para evitar el sobreajuste. Obtiene el menor RMSE, lo que sugiere un buen rendimiento en la predicción.
- La Red Neuronal 2 utiliza la función de activación Tanh y L2 regularization. Aunque tiene un RMSE ligeramente mayor, puede estar capturando relaciones más sutiles entre las variables.
- La Red Neuronal 3 utiliza Leaky ReLU y Batch normalization. Obtiene un rendimiento similar a la Red Neuronal 1, pero puede converger más rápido durante el entrenamiento debido a la normalización.

**4. Selección de la red neuronal óptima:**

La Red Neuronal 1 parece ser la opción óptima, ya que logra el menor RMSE en el conjunto de prueba y utiliza una combinación efectiva de funciones de activación y técnicas de regularización para evitar el sobreajuste.

Basándonos en los resultados de RMSE en el conjunto de prueba que proporcionaste previamente, podemos determinar cuál de las tres redes neuronales es la más óptima en función de su rendimiento:

- Red Neuronal 1 (ReLU y Dropout): RMSE en conjunto de prueba: 0.15
- Red Neuronal 2 (Tanh y L2 Regularization): RMSE en conjunto de prueba: 0.17
- Red Neuronal 3 (Leaky ReLU y Batch Normalization): RMSE en conjunto de prueba: 0.16

La Red Neuronal 1 con función de activación ReLU y Dropout tiene el RMSE más bajo en el conjunto de prueba (0.15). Esto significa que es la que mejor predice la aprobación (approval_Index) de las películas en función de las variables disponibles en el dataset, al menos según los resultados de evaluación que proporcionaste previamente.

Por lo tanto, en base a los resultados, la Red Neuronal 1 parece ser la opción más óptima para este problema de predicción de aprobación de películas.